legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabTokenManager.java

```diff
@@ -175,27 +191,41 @@ void putAllFromToken(Token.TokenReader reader)
                 // Read values
                 String modeName = reader.getString();
                 String appId = reader.getString();
+                String typeName = reader.getString();
                 String token = reader.getString();
 
-                if ((modeName != null) && (appId != null) && (token != null))
+                if ((modeName != null) && (appId != null) && (typeName != null) && (token != null))
                 {
-                    // Get mode
+                    // Get mode and token type
                     GitLabMode mode;
+                    TokenType type;
                     try
                     {
                         mode = GitLabMode.valueOf(modeName);
+                        type = TokenType.valueOf(typeName);
                     }
                     catch (IllegalArgumentException e)
                     {
-                        // unknown mode - token will be ignored
+                        // unknown mode or token type - token will be ignored
                         continue;
                     }
 
                     // Check the mode info
                     GitLabModeInfo modeInfo = this.modeInfos.getModeInfo(mode);
                     if ((modeInfo != null) && appId.equals(modeInfo.getAppInfo().getAppId()))
                     {
-                        this.tokens.put(mode, token);
+                        if (type.equals(TokenType.OAUTH2_ACCESS))
+                        {
+                            this.tokens.put(mode, GitLabToken.newOAuthToken(token));
+                        }
+                        else if (type.equals(TokenType.PRIVATE))
+                        {
+                            this.tokens.put(mode, GitLabToken.newPrivateAccessToken(token));
+                        }
+                        else
+                        {
+                            throw new UnsupportedOperationException(typeName + " token is not supported");
+                        }
```

Comment:
If we add a factory method on GitLabToken that takes both TokenType and String, we can eliminate this logic.